## Final Project Submission


* Student name: Annika Noren
* Student pace: part time 
* Scheduled project review date/time: 
* Instructor name: Abhineet Kulkarni
* Blog post URL:


# Exploratory Data Analysis
I will use basic python commands to load, view and learn about the data set, Kings' County Housing Data.

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv("kc_house_data.csv")

In [2]:
df.head()

id        date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014  221900.0         3       1.00         1180   
1  6414100192   12/9/2014  538000.0         3       2.25         2570   
2  5631500400   2/25/2015  180000.0         2       1.00          770   
3  2487200875   12/9/2014  604000.0         4       3.00         1960   
4  1954400510   2/18/2015  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view     ...      grade  sqft_above  \
0      5650     1.0         NaN   0.0     ...          7        1180   
1      7242     2.0         0.0   0.0     ...          7        2170   
2     10000     1.0         0.0   0.0     ...          6         770   
3      5000     1.0         0.0   0.0     ...          7        1050   
4      8080     1.0         0.0   0.0     ...          8        1680   

   sqft_basement yr_built  yr_renovated  zipcode      lat     long  \
0            0.0     1955           0.0    98178  47.5112 -122.257   
1          400.0     1951        1991.0    98125  47.7210 -122.319   
2            0.0     1933           NaN    98028  47.7379 -122.233   
3          910.0     1965           0.0    98136  47.5208 -122.393   
4            0.0     1987           0.0    98074  47.6168 -122.045   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503  

[5 rows x 21 columns]

In [3]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.159700e+04  2.159700e+04  21597.000000  21597.000000  21597.000000   
mean   4.580474e+09  5.402966e+05      3.373200      2.115826   2080.321850   
std    2.876736e+09  3.673681e+05      0.926299      0.768984    918.106125   
min    1.000102e+06  7.800000e+04      1.000000      0.500000    370.000000   
25%    2.123049e+09  3.220000e+05      3.000000      1.750000   1430.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.159700e+04  21597.000000  19221.000000  21534.000000  21597.000000   
mean   1.509941e+04      1.494096      0.007596      0.233863      3.409825   
std    4.141264e+04      0.539683      0.086825      0.765686      0.650546   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068500e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above      yr_built  yr_renovated       zipcode  \
count  21597.000000  21597.000000  21597.000000  17755.000000  21597.000000   
mean       7.657915   1788.596842   1970.999676     83.636778  98077.951845   
std        1.173200    827.759761     29.375234    399.946414     53.513072   
min        3.000000    370.000000   1900.000000      0.000000  98001.000000   
25%        7.000000   1190.000000   1951.000000      0.000000  98033.000000   
50%        7.000000   1560.000000   1975.000000      0.000000  98065.000000   
75%        8.000000   2210.000000   1997.000000      0.000000  98118.000000   
max       13.000000   9410.000000   2015.000000   2015.000000  98199.000000   

                lat          long  sqft_living15     sqft_lot15  
count  21597.000000  21597.000000   21597.000000   21597.000000  
mean      47.560093   -122.213982    1986.620318   12758.283512  
std        0.138552      0.140724     685.230472   27274.441950  
min       47.155900   -122.519000     399.000000     651.000000  
25%       47.471100   -122.328000    1490.000000    5100.000000  
50%       47.571800   -122.231000    1840.000000    7620.000000  
75%       47.678000   -122.125000    2360.000000   10083.000000  
max       47.777600   -121.315000    6210.000000  871200.000000

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
id               21597 non-null int64
date             21597 non-null object
price            21597 non-null float64
bedrooms         21597 non-null int64
bathrooms        21597 non-null float64
sqft_living      21597 non-null int64
sqft_lot         21597 non-null int64
floors           21597 non-null float64
waterfront       19221 non-null float64
view             21534 non-null float64
condition        21597 non-null int64
grade            21597 non-null int64
sqft_above       21597 non-null int64
sqft_basement    21597 non-null object
yr_built         21597 non-null int64
yr_renovated     17755 non-null float64
zipcode          21597 non-null int64
lat              21597 non-null float64
long             21597 non-null float64
sqft_living15    21597 non-null int64
sqft_lot15       21597 non-null int64
dtypes: float64(8), int64(11), object(2)
memory usage: 3.5+ MB


In [5]:
print(df.shape)
print(df.isna().sum())

(21597, 21)
id                  0
date                0
price               0
bedrooms            0
bathrooms           0
sqft_living         0
sqft_lot            0
floors              0
waterfront       2376
view               63
condition           0
grade               0
sqft_above          0
sqft_basement       0
yr_built            0
yr_renovated     3842
zipcode             0
lat                 0
long                0
sqft_living15       0
sqft_lot15          0
dtype: int64


In [6]:
# Column Names and descriptions for Kings County Data Set
# * **id** - unique identified for a house
# * **date** - house was sold
# * **price** -  is prediction target
# * **bedrooms** -  of bedrooms
# * **bathrooms** -  of bathrooms
# * **sqft_living** -  footage of the home
# * **sqft_lots** -  footage of the lot
# * **floors** -  floors (levels) in house
# * **waterfront** - House which has a view to a waterfront
# * **view** - Has been viewed
# * **condition** - How good the condition is ( Overall )
# * **grade** - overall grade given to the housing unit, based on King County grading system
# * **sqft_above** - square footage of house apart from basement
# * **sqft_basement** - square footage of the basement
# * **yr_built** - Built Year
# * **yr_renovated** - Year when house was renovated
# * **zipcode** - zip
# * **lat** - Latitude coordinate
# * **long** - Longitude coordinate
# * **sqft_living15** - The square footage of interior housing living space for the nearest 15 neighbors
# * **sqft_lot15** - The square footage of the land lots of the nearest 15 neighbors


In [7]:
import matplotlib.pyplot as plt
fig = plt.figure(figsize = (12,12))
ax = fig.gca()
df.hist(ax = ax);

/Users/Annika/anaconda3/envs/learn-env/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2961: UserWarning: To output multiple subplots, the figure containing the passed axes is being cleared
  exec(code_obj, self.user_global_ns, self.user_ns)


### Summary of EDA
It is a large data set, with 21597 rows of information and 21 columns.  There appears to be some missing data and there may be some duplicate information, which needs to be investigated.  Based on the histrograms, some features may be categorical; it will be a good idea to make sure the categories are proper before starting linear regression.  Additionally, some features don't seem to make sense when predicting sales price so they will be looked at.

# Dealing with Extraneous Data
There are three features - sqft_living, sqft_above and sqft_basement - that may be redundant.  The sum of above and basement equals living.   Therefore, I will reduce the columns by dropping those two. 

In [8]:
print(df['sqft_basement'].nunique())
print(df.sqft_basement.value_counts().head())

304
0.0      12826
?          454
600.0      217
500.0      209
700.0      208
Name: sqft_basement, dtype: int64


In [9]:
del df['sqft_basement']
del df['sqft_above']

In [10]:
df.head()

id        date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014  221900.0         3       1.00         1180   
1  6414100192   12/9/2014  538000.0         3       2.25         2570   
2  5631500400   2/25/2015  180000.0         2       1.00          770   
3  2487200875   12/9/2014  604000.0         4       3.00         1960   
4  1954400510   2/18/2015  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  condition  grade  yr_built  \
0      5650     1.0         NaN   0.0          3      7      1955   
1      7242     2.0         0.0   0.0          3      7      1951   
2     10000     1.0         0.0   0.0          3      6      1933   
3      5000     1.0         0.0   0.0          5      7      1965   
4      8080     1.0         0.0   0.0          3      8      1987   

   yr_renovated  zipcode      lat     long  sqft_living15  sqft_lot15  
0           0.0    98178  47.5112 -122.257           1340        5650  
1        1991.0    98125  47.7210 -122.319           1690        7639  
2           NaN    98028  47.7379 -122.233           2720        8062  
3           0.0    98136  47.5208 -122.393           1360        5000  
4           0.0    98074  47.6168 -122.045           1800        7503

Additionally, the features of zipcode, latitude and longitude are redundant.  Therefore, I will reduce the data set further be deleting lat and long.

In [11]:
del df['lat']
del df['long']

In [12]:
df.head()

id        date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014  221900.0         3       1.00         1180   
1  6414100192   12/9/2014  538000.0         3       2.25         2570   
2  5631500400   2/25/2015  180000.0         2       1.00          770   
3  2487200875   12/9/2014  604000.0         4       3.00         1960   
4  1954400510   2/18/2015  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  condition  grade  yr_built  \
0      5650     1.0         NaN   0.0          3      7      1955   
1      7242     2.0         0.0   0.0          3      7      1951   
2     10000     1.0         0.0   0.0          3      6      1933   
3      5000     1.0         0.0   0.0          5      7      1965   
4      8080     1.0         0.0   0.0          3      8      1987   

   yr_renovated  zipcode  sqft_living15  sqft_lot15  
0           0.0    98178           1340        5650  
1        1991.0    98125           1690        7639  
2           NaN    98028           2720        8062  
3           0.0    98136           1360        5000  
4           0.0    98074           1800        7503

# Dealing with Missing and Duplicate Data
There are 3 features that are missing data: View (how many times the property has been viewed), Waterfront (is it a water front property or not) and Year Renovated.  There are also duplicate entries of properties with the same id.

### Checking for unique property id

In [13]:
print(df['id'].nunique()) #there could be duplicated data
print('----')
print(df['id'].value_counts().head(3)) #there is one triplicate data, the others are duplicate

21420
----
795000620     3
1825069031    2
2019200220    2
Name: id, dtype: int64


In [14]:
duplicates = df[df.duplicated(subset='id')] 
print('There are {} occurances of duplicate house id numbers.'.format(len(duplicates))) #177 occurances of 21597
percent = round(len(duplicates) / len(df), 3) * 100
print('Percent of data that is duplicate: ', percent)

There are 177 occurances of duplicate house id numbers.
Percent of data that is duplicate:  0.8


In [17]:
#The duplicate house ids will be deleted
df = df.drop_duplicates(['id'], keep='last') #deleted the duplicate rows
# print(df['id'].value_counts().head(3))
# print(df['id'].nunique())
print(df.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21420 entries, 0 to 21596
Data columns (total 17 columns):
id               21420 non-null int64
date             21420 non-null object
price            21420 non-null float64
bedrooms         21420 non-null int64
bathrooms        21420 non-null float64
sqft_living      21420 non-null int64
sqft_lot         21420 non-null int64
floors           21420 non-null float64
waterfront       19067 non-null float64
view             21357 non-null float64
condition        21420 non-null int64
grade            21420 non-null int64
yr_built         21420 non-null int64
yr_renovated     17607 non-null float64
zipcode          21420 non-null int64
sqft_living15    21420 non-null int64
sqft_lot15       21420 non-null int64
dtypes: float64(6), int64(10), object(1)
memory usage: 2.9+ MB
None


### Checking values for Waterfront,  View and Yr_Renovated

In [18]:
print(df['waterfront'].value_counts()) #0 is view, Nan not a view?
print(df['waterfront'].isna().sum())# 2353 NaN values, 146 as "1" and 189921 as "0" but don't know what that means

0.0    18921
1.0      146
Name: waterfront, dtype: int64
2353


There are 18,921 properties with a '0' for waterfront, 146 with a '1' and 2,353 with no value. Since Waterfront is a categorical column and there isn't key to identify what 1 and 0 represent, I'll leave the missing values for now and revisit if it seems to be a problem.  Or assume 0 means no view and high percentage have no view, so it is feature to ignore.

In [19]:
#print(df['view'].nunique())#5
print(df['view'].value_counts())
print(df['view'].isna().sum())

0.0    19253
2.0      956
3.0      505
1.0      329
4.0      314
Name: view, dtype: int64
63


There are 5 categories for the view feature with 63 null values.  Since this is a categorical column and we do not know the true meaning of the assigned category (assumptions can be made), I'll leave the missing values for now and revisit if it seems to be a problem.

In [20]:
print(df['yr_renovated'].nunique())
print(df['yr_renovated'].isna().sum())

70
3813


There are 3,813 missing values.  In place of making up data, I will replace the data with "2020" - a nonsense value.

In [ ]:
df['yr_renovated'] = df['yr_renovated'].fillna(value='2020')
df['yr_renovated'].isna().sum()

## Result of Cleaning 

In [ ]:
df.info()

### Identify Categorial Data
I will use scatterplots to identify the categorial features. Some features do not represent a continues number, but a class or state.  

In [ ]:
fig, axes = plt.subplots(nrows=1, ncols=3, figsize=(16,3))

for xcol, ax in zip(['bathrooms', 'bedrooms','floors'], axes):
    df.plot(kind='scatter', x=xcol, y='price', ax=ax, alpha=0.4, color='b')    

In [ ]:
fig, axes = plt.subplots(nrows=1, ncols=3, figsize=(16,3))

for xcol, ax in zip(['lat','long', 'zipcode'], axes):
    df.plot(kind='scatter', x=xcol, y='price', ax=ax, alpha=0.4, color='b')  

In [ ]:
fig, axes = plt.subplots(nrows=1, ncols=4, figsize=(16,3))

for xcol, ax in zip(['grade', 'condition','sqft_living15','sqft_lot15'], axes):
    df.plot(kind='scatter', x=xcol, y='price', ax=ax, alpha=0.4, color='b') 

In [ ]:
fig, axes = plt.subplots(nrows=1, ncols=3, figsize=(16,3))

for xcol, ax in zip(['view', 'waterfront','yr_built'], axes):
    df.plot(kind='scatter', x=xcol, y='price', ax=ax, alpha=0.4, color='b') 

In [ ]:
fig, axes = plt.subplots(nrows=1, ncols=3, figsize=(16,3))

for xcol, ax in zip(['sqft_above','sqft_living', 'sqft_lot'], axes):
    df.plot(kind='scatter', x=xcol, y='price', ax=ax, alpha=0.4, color='b') 

### Creating dummy variables for View, Waterfront, Grade and Condition 


In [ ]:
# view_dummy = pd.get_dummies(df["view"], prefix="view")
# waterfront_dummy = pd.get_dummies(df["waterfront"], prefix="w_fr")
# grade_dummy = pd.get_dummies(df["grade"], prefix="grade")
# condition_dummy = pd.get_dummies(df["condition"], prefix="cond")
#yr_built_dummies = pd.get_dummies(df["yr_built"], prefix="yr_b")

In [ ]:
# df_categories = pd.concat([view_dummy,waterfront_dummy, grade_dummy,condition_dummy], axis=1)
# df_categories.head()

In [ ]:
#drop old columns and concat with new columns
# df = df.drop(['view','waterfront','grade','condition'], axis=1)
# df = pd.concat([df,view_dummy, waterfront_dummy,grade_dummy,condition_dummy], axis=1)
# df.head()

### Dealing with yr_built
Since there are so many values for years that properties were built, instead of creating dummy variables it may be better to create bins for the data. IM SECOND GUESSING MY NEED TO BIN THIS DATA

In [ ]:
print(df["yr_built"].nunique())
#print(df.yr_built.value_counts()) 

In [ ]:
# bins = [1901, 1940, 1955, 1970, 1985, 2000, 2020]
# bins_yr_built = pd.cut(df['yr_built'], bins)
# bins_yr_built = bins_yr_built.cat.as_ordered()
# bins_yr_built.head()

In [ ]:
# bins_yr_built.value_counts().plot(kind='bar')

In [ ]:
# yr_built_dummy = pd.get_dummies(bins_yr_built, prefix="yr_b")
# # df["yr_built"]=bins_yr_built
# # df.iloc[0:5,9:20]
# df = df.drop(['yr_built'], axis=1)
# df = pd.concat([df,yr_built_dummy], axis=1)
# df.head()

### Reducing the size of the dataset
I'm going to delete some columns that I think are useless so I get move forward with a correlation

In [ ]:
#df_features = df[['bedrooms', 'bathrooms', 'sqft_living','sqft_lot', 'floors']].copy()

In [ ]:
#pd.plotting.scatter_matrix(df,figsize  = [12,12]);